### ISSUE TITLE
[OTBN/SW] Cause a bus error on access to OTBN memories whilst OTBN is active

### ISSUE THREAD

**Issue by GregAC** [lowRISC/opentitan#2696]
The initial OTBN implementation blocks any access to the OTBN I and D mem whilst the OTBN core is busy. This seems like a sensible design but has the side-effect that if Ibex attempts to access either of these memories whilst the OTBN core is busy it will stall until OTBN finishes.

I propose any access to the OTBN memories whilst it is running should result in a bus error. It seems any attempt by software to access the memories whilst it is running would be a bug and it is preferable that results in an exception rather than Ibex stalling for an indeterminate period. If there's a bug in the OTBN software then it may never terminate and the whole system would lock up.

> estimate 4
> remaining 2021-08-09 2

**Comment by lenary** [comment]
Yeah, a bus error sounds better than silently dropping the request or stalling (potentially forever).

I don't know if eventually we can set up a PMP region so you end up with a PMP error rather than a bus error, but that can come later.

**Comment by tjaychen** [comment]
do you really want to error on all requests though? I know OTBN right now is interrupt driven, but it's conceivable the host processor may want to poll on some status or condition for the sake of some application or software work-around.

The way i've been thinking about it is that the internals of OTBN should not be modifiable while it is ongoing, but it should still be readable (at least portions of it).

If you wanted to accomplish that function, i think it would be better to build in some kind of gaskets that error on writes, but allow reads to go through while OTBN is non-idle.  The same could be applied to the entire register space. 

This same concept is currently applied to the clock manager and how it manages the OTBN clock. (Or at least will be applied).

**Comment by rswarbrick** [comment]
That's a good point but would it be easier just to have some status registers, rather than a shared memory? That seems much easier to audit from a security point of view and if all you need is "have I finished phase N?", it's more than enough.

**Comment by GregAC** [comment]
I agree with @rswarbrick having some seperate space that can be used for OTBN to report status to that Ibex can read whilst it's running seems like a useful feature. Maybe just an extra set of CSRs that are visible to the memory system (either read-only or read-write, maybe these access permissions could be controlled by OTBN).

**Comment by lenary** [comment]
Yeah, this does seem like the right compromise:
- Bus error on trying to access d/imem while OTBN is running.
- A few specific registers for signalling information between OTBN and the main ibex core (My view is these should probably be uni-directional for security, but have more than one for each direction).

**Comment by imphil** [comment]
A status register already exists and can be used for polling. If desired, we can extend this status register to provide more fine-grained status (i.e. status within the run-to-completion operation). That hasn't been a requirement so far and would be a separate issue anyway.

The protocol how to use OTBN is as follows:
- Access IMEM and DMEM as you wish.
- Start the operation by writing the START bit.
- Now accesses to IMEM and DMEM are forbidden.
- Either: wait for the done interrupt, or poll the status register (or both to be sure)
- Now IMEM and DMEM can be accessed again.

The question in this issue is only relevant if this protocol is violated.

**Comment by tjaychen** [comment]
sounds good. If there is already status registers or plan for more information, then i agree the memory should just error back. I'm still a bit hesitant to say that it should not even be readable, since I have seen ugly software work-arounds that relied on being able to poll memory, but hopefully that's not something that should ever come.

**Comment by moidx** [comment]
From a programming model perspective we are planning to rely on the done interrupt and the status register to manage execution. 

One thing we have not added to the programming model yet is the ability to reset OTBN from Ibex. This may be used to simplify OTBN state management. Updating @imphil's flow (from ibex perspective):

* mutex lock
* Initialize OTBN by applying reset + secure initialization (scrambling, etc) 
* Access IMEM and DMEM as you wish.
* Start the operation by writing the START bit.
* Now accesses to IMEM and DMEM are forbidden.
* Either: wait for the done interrupt, or poll the status register (or both to be sure)
* Now IMEM and DMEM can be accessed again.
* cleanup
* mutex unlock 

The reset can also be used to recover from any stalls in OTBN, although stalls at the moment are considered programming errors.

The ability to communicate a more fine grained status between OTBN and Ibex is desired, either via a status register or by allowing access to DMEM. 

Some of the potential implementation patterns that @tjaychen is referring to in an earlier comment, also require the ability for OTBN to poll for an update provide by Ibex. Here again we can support this either via a register or by allowing access to DMEM. 

At first, I think we are considering denying access to DMEM to get better isolation, but it is also important to note that Ibex and OTBN are still inside the same trusted compute base. We are planning to integrity protect both components with the same key, and Ibex is in charge of configuring OTBN. This should be taken into consideration when determining whether to allow access to DMEM during OTBN execution.

**Comment by imphil** [comment]
KMAC had the same design as we have and how changed it to not hang the bus, but allow the access and set an error code register instead: https://github.com/lowRISC/opentitan/pull/4412 (issue: https://github.com/lowRISC/opentitan/issues/4409)

We probably should just do the same in OTBN?

**Comment by tjaychen** [comment]
yeah i agree with Philip.
I think an access while active more likely will end up being a software
mistake, so if we just hang the system it'll make things difficult to
debug.

**Comment by GregAC** [comment]
So to confirm we now want a read to OTBN memory from Ibex when OTBN is running to not signal a bus error just set an error code register?

We'd need a new error code register as otherwise you have to deal with interactions with the OTBN error code register.

This also means that any time software reads back from OTBN memory it must check this error code register after the read is done to ensure it was successful.

I note kmac also raises its error interrupt when this happens, which OTBN doesn't have.

**Comment by rswarbrick** [comment]
I don't think I understand. Why don't we just respond with a bus error? Seems much simpler for SW to deal with than a spooky interrupt coming in from around the back.

**Comment by GregAC** [comment]
I'd agree, I think the bus error is a better option.

**Comment by lenary** [comment]
Yeah, from the SW side, a Bus Error is *far* easier than having to check an extra register in a lot of extra places.

**Comment by tjaychen** [comment]
i think what @imphil and @moidx original point was that as part of the `obtn` programmer's flow, we must look at the status after a completion anyways to check if anything nefarious has happened.  So it's reasonable to lump in "i was accessed when i should not have been" into that category.

So this i think is a bit different from the desire to have a bus error.  That I understand to be more like a software debugging tool than it is a general status check.

So my view is..

1. we definitely should not hang the bus (i think we all agree)
2. we should also definitely include it as a status bit / error code (this could be especially useful in a world where we have multi-hosts and one of them was made to do something undesirable)
3. I think having a bus error is also fine, but we should follow the guidelines shown [here](https://docs.opentitan.org/doc/rm/register_tool/#error-responses).  Exactly what `devmode_i` will be hooked up to I think is still somewhat open, but we should reserve the option to remove software debuggability in certain situations.

**Comment by rswarbrick** [comment]
Hmm, isn't the "return garbage" behaviour to avoid people probing the address map? In this case, we're not hiding the size of memory (we're just stopping it from working some of the time) so I don't really see the point. Or is there some other reason to do this?

**Comment by GregAC** [comment]
Having the bus error may be helpful to attackers trying to craft an glitch attack, they know their glitch has done something interesting so the glitch that helps them advance their attack may be close by.

If we don't want a bus error and want this to feed into the OTBN status then we need to consider how it interacts with other OTBN errors, I see three options:

1. Halt OTBN immediately on the improper access, this raises the done interrupt and sets the appropriate error code, exactly like any other OTBN error
2. Leave OTBN running, note there has been an improper access somewhere, when OTBN halts if there's been some other (internal to OTBN) error that takes priority as the error code, otherwise set 'bad bus access' as error code
3. Add a new field to the error  code register or an entirely new error code register for bad bus accesses and anything else that might fall into this category (error whilst OTBN running but not related to OTBN's execution).

**Comment by tjaychen** [comment]
would be good to hear what @imphil, @moidx and @cdgori think.
But I tend to favor leaving OTBN running and register the fact that something improper has happened. 
That way from a macro perspective, the overall timing does not change. 

I'm a bit behind on the error code, are you saying that at the moment only one error is registered upon completion? I think it would be good to do as you mention in point 3 and register the errors in different fields and leave the decision to software later.  Sorry I dont know if you guys have already discussed this separately and settled on single field already.  If that's the case, ignore what I'm saying here.

**Comment by tjaychen** [comment]
back to your point though Greg, I think having a bus error can be very useful.  It's just that it'll be more of a development function in my view.

**Comment by cdgori** [comment]
For development I understand having the bus error as a convenience, but in deployed software it should never ever happen, right?

Assuming that is correct, then we have to consider @GregAC 's 3 options.  Option 1 seems to be problematic because it potentially gives (immediate) information about the success of a glitch, so I prefer option 2 (I could see option 3 as well, but using the existing mechanisms is a bit simpler, I would hope.)

Question about the phrasing of option 1, is it correct that for all/most discovered errors that OTBN immediately halts?  This is obviously a bit of a tricky thing to sort out, since some classes of errors you don't want to proceed, but some you want to minimize the direct feedback to the adversary about their success in glitching, and which is which is something of a judgment call.

**Comment by GregAC** [comment]
@cdgori yes current behaviour is immediate halt, raising the `done` interrupt on any error.

Aside for ECC faults all of the current errors indicate something illegal/undefined happening with the instructions so whilst we could ignore them and continue there's no guarantee the software would terminate (or would terminate far later/earlier than it normally would).

We could use a software timeout based on estimated completion time. Then OTBN could continue past errors and SW would timeout if it doesn't get to `done` when expected. We'd need to add functionality for SW to forcefully halt OTBN to do this.

It would be straight-forward to categorise errors into immediately halt vs remember and continue if desired.

**Comment by rswarbrick** [comment]
The current behaviour is that OTBN stops immediately if something goes wrong *in OTBN*, mainly because the processor doesn't really know what else to do. I believe that we've discussed this several times (including in security meetings), and agreed that approaches like "hey, let's run for a bit more and see what happens" are a terrible idea. (One thing that might happen is that the glitched software leaks key data somewhere handy...). If we want to talk about that design decision again, let's do it in a separate issue.

Going back to this issue, the discussion is about what we should do on a bus access from the host processor when OTBN is running. In this case, OTBN could choose to pootle on as if nothing has happened (options 2 or 3 above). Or it could come to a screaming halt (option 1).

As Chris said, option 1 doesn't make much sense, since doing it would give the attacker the same amount of information as a bus error. I also really dislike "option 2" as written, because it's a bit confusing. You might get told about the bad access, or you might not, depending on other things. Eugh! That leaves "option 3", but I'm worried that it might make SW error detection a bit more fiddly.

I propose we do the following:

- On a memory access from the host processor when OTBN is running, stores are acknowledged (and dropped) and loads always return `0` (we could use random data, but does that really help? Or, if we're thinking about power signatures, we could work on the basis that most data is "half-and-half" and return `0xaaaaaaaa`)
- In either case, an internal flag is set.
- We change OTBN's error status to be a bitfield. We've got 32 bits and far fewer than 32 errors at the moment, so this should be fine. If we need more in the future, we can use a "top bit chaining" mechanism (bit 31 means that something else went wrong; look at register $FOO for details)
- When OTBN finishes, the internal flag is reflected as a bit in the error bitfield.

This way:
- Host software can check for success by comparing the status with zero.
- An attacker doesn't get any immediate feedback to show the glitch was successful.

Are there any other features that we want that aren't addressed by this proposal?

**Comment by GregAC** [comment]
Would we switch to all errors having their own bit or retain the current set of error codes in the bottom 8 bits and introduce a new field in the same register that could be a bitfield (one of which will be set if we have the invalid access from Ibex)? I was thinking we'd retain the existing error codes and add a separate field as one of the ways to do option 3.

**Comment by rswarbrick** [comment]
Personally, I prefer the uniformity of using a bitfield for everything.

**Comment by tjaychen** [comment]
> The current behaviour is that OTBN stops immediately if something goes wrong _in OTBN_, mainly because the processor doesn't really know what else to do. I believe that we've discussed this several times (including in security meetings), and agreed that approaches like "hey, let's run for a bit more and see what happens" are a terrible idea. (One thing that might happen is that the glitched software leaks key data somewhere handy...). If we want to talk about that design decision again, let's do it in a separate issue.
> 
> Going back to this issue, the discussion is about what we should do on a bus access from the host processor when OTBN is running. In this case, OTBN could choose to pootle on as if nothing has happened (options 2 or 3 above). Or it could come to a screaming halt (option 1).
> 
> As Chris said, option 1 doesn't make much sense, since doing it would give the attacker the same amount of information as a bus error. I also really dislike "option 2" as written, because it's a bit confusing. You might get told about the bad access, or you might not, depending on other things. Eugh! That leaves "option 3", but I'm worried that it might make SW error detection a bit more fiddly.
> 
> I propose we do the following:
> 
> * On a memory access from the host processor when OTBN is running, stores are acknowledged (and dropped) and loads always return `0` (we could use random data, but does that really help? Or, if we're thinking about power signatures, we could work on the basis that most data is "half-and-half" and return `0xaaaaaaaa`)
> * In either case, an internal flag is set.
> * We change OTBN's error status to be a bitfield. We've got 32 bits and far fewer than 32 errors at the moment, so this should be fine. If we need more in the future, we can use a "top bit chaining" mechanism (bit 31 means that something else went wrong; look at register $FOO for details)
> * When OTBN finishes, the internal flag is reflected as a bit in the error bitfield.
> 
> This way:
> 
> * Host software can check for success by comparing the status with zero.
> * An attacker doesn't get any immediate feedback to show the glitch was successful.
> 
> Are there any other features that we want that aren't addressed by this proposal?

@rswarbrick 
yes i really like this approach of separating out internal errors and external harmless errors that might be caused by the system.
I also agree with your suggestion of keep this as status and for the host to check it later.

Regarding data return, I also feel a fixed value should be fine, with a slight bias to something "random".  Mostly because if we return the same thing all the time, I'm not sure if that will end up looking a specific way on the power trace and basically give an attacker information about when to mount an attack.  I do not think this is a huge concern though.  Do you have any thoughts Chris? 

Lastly, do you still want to reserve an option for bus errors under certain conditions?  That would still be helpful for software development as long as we can disable it (we can talk about the disable separately, I have some thoughts here because there are 2 other places in the system that have the same issue).

**Comment by rswarbrick** [comment]
I'm not sure about the bus errors for development. It kind of feels like SW authors can probably cope with a "don't do that...", but maybe there's something I'm missing. @lenary, do you have an opinion?

For Sam, to avoid having to read the entire thread: This discussion is about what happens if the host processor makes a bus access to OTBN's memories when OTBN is running. We don't want to allow any information leakage in this case. In production, we've agreed that writes will be dropped and reads will get some arbitrary value. The question is whether we want to respond with a bus error to help software bringup when in test mode. Do you think that would make much difference?

**Comment by tjaychen** [comment]
It's probably worth adding that during production, the fact that there was
a processor access may be registered in the status.
So I was thinking during development we would want a way to remove these as
they pop up, and a bus error would be a quick way to do this.
But to Rupert's point, this definitely can be handled in other ways.

**Comment by cdgori** [comment]
No strong feeling about the read return value (0x0 vs 0xAAAA...) - I doubt it makes much difference.

I actually still do (slightly) prefer option 2.  The SW is only not going to be told about a bad access if something far, far more serious has happened, and this approach adds no more state/hardware.  In theory there might (?) be value in having some "error priority" such that more serious errors take priority over less serious ones, just in terms of limiting information to a would-be attacker.  (This is fully at odds with convenience to the SW team in dev/bring-up, of course.)

However, if you want to do the bitfield-of-errors I can accept that, I think.  But, then I would have some slight preference to collapsing the bitfield to a single pass/fail bit in PROD and allowing access to the full bitfield in DEV.  This goes back to my point that these errors are not expected to occur in normal operation as a result of software actions but rather attacks, so not providing detailed status is preferable.

Basically, I'm willing to make debug in PROD quite difficult/challenging/opaque, because debugging is supposed to happen in DEV (or RMA in some cases).

**Comment by rswarbrick** [comment]
I think I agree about collapsing errors in production mode. That raises the obvious question about how OTBN figures out the lifecycle state. Let's track that separately at #4519.

**Comment by tjaychen** [comment]
hey all,
i'll set up some time for us to discuss next week.
I just want to caution though that in my experience, typically software
doesn't like a status changing on them depending on the state of the chip.
I think it makes sense to not assert certain errors if that's what we want,
but it's weird to say "this field is now invalid etc".  It puts a bit of an
extra burden on software to remember what to track.  (of course this can be
pretty easily wrapped up in DIFs)

I should also caution that IMO what will happen in production is that
development parts may end up being hand crafted because of how special they
are. Which means I think for large parts of development software will
probably be using production parts.
So if debug is really painful in that scenario, it might become really
unwieldy.  Especially if the development environment is heavily integrated
with a wider system, which might make FPGAs harder to use also.

**Comment by GregAC** [comment]
> typically software
doesn't like a status changing on them depending on the state of the chip.

If we do change between some detailed status and a simple pass/fail then this may not be a huge issue. It both cases we can define things such that != 0 means fail/error. Maybe even introduce due a specific code/bit field that is a 'generic error' which we use in PROD for everything and only use the more specific codes in DEV/RMA.

**Comment by rswarbrick** [comment]
Yep, that's the sort of thing I was imagining too. We're having a meeting tomorrow evening (UK time): let's thrash out the details there.

**Comment by GregAC** [comment]
After a meeting to discuss we've chosen something close to option 1.

On an external access to OTBN errors when OTBN is active, we will:

* Raise a fatal alert
* Bring OTBN to an immediate halt (raising the `done` interrrupt if it's enabled)
* Set an appropriate bit in the OTBN error bitfield (see https://github.com/lowRISC/opentitan/issues/4597)
